docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+
+The idea is introduce a **consistency check in the querier**. When a store-gateway receives a request from the querier, the store-gateway includes in the response the list of block IDs currently loaded on the store-gateway itself. The querier can then merge the list of block IDs received from all store-gateway hit, and match it against the list of block IDs computed at the beginning of the query execution.
+
+There are three possible scenarios:
+
+1. The list match: all good
+2. All the blocks known by the querier are within the list of blocks returned by store-gateway, but the store-gateway also included blocks unknown to the querier: all good (it means the store-gateways have discovered and loaded new blocks before the querier discovered them)
+3. Some blocks known by the querier are **not** within the list of blocks returned by store-gateway: potential consistency issue
+
+We want to protect from a partial results response which may occur in the case #3. However, there are some legit cases which, if not handled, would lead to frequent false positives. Given the querier and store-gateway instances independently scan the bucket at a regular interval (to find new blocks or deleted blocks), we may be in one of the following cases:
+
+a. The querier has discovered new blocks before the store-gateway successfully discovered and loaded them
+b. The store-gateway has offloaded blocks "marked for deletion" before the querier
+
+To protect from case (a), we can exclude the blocks which have been uploaded in the last `X` time from the consistency check (same technique already used in other Thanos components). This `X` delay time is used to give the store-gateway enough time to discover and load new blocks, before the querier consider them for the consistency check. This value `X` should be greater than the `-experimental.tsdb.bucket-store.consistency-delay`, because we do expect the querier to consider a block for consistency check once it's reasonably safe to assume that its store-gateway already loaded it.
+
+To protect from case (b) we need to understand how blocks are offloaded. The `BucketStore` (running within the store-gateway) offloads a block as soon as it's not returned by the `MetaFetcher`. This means we can configure the `MetaFetcher` with a [`IgnoreDeletionMarkFilter`](https://github.com/thanos-io/thanos/blob/4bd19b16a752e9ceb1836c21d4156bdeb517fe50/pkg/block/fetcher.go#L648) with a delay of `X` (could be the same value used for case (a)) and in the querier exclude the blocks which have been marked for deletion more than `X` time ago from the consistency check.
+
+## Trade-offs
+
+The proposed solution comes with the following trade-offs:
+
+- A querier is not ready until it has completed an initial full scan of the bucket, downloading the `meta.json` file of every block
```

Comment:
All of those we can fix in later steps with our old proposal of different block ID right @pracucci ?

Reply:
Not sure. According to this proposal we would need the "marked for deletion" information as well, so even if the min/max timestamp is in the storage prefix, it may not be enough.

Reply:
good point. But at least it will improve a lot.